### PR TITLE
jormungandr: feed_publisher in distributed for here and geovelo

### PR DESCRIPTION
**Pull request by kadhikari**
Ticket: https://jira.canaltp.fr/browse/NAVP-761

### CODE REVIEW

File: source/jormungandr/tests/here_distributed_routing_tests.py [hunk 1]
```diff
@@ -347,3 +347,16 @@ def test_basic_car_routing(self):
...
         assert sections[2].get('mode') == 'car'
+
+        feeds = get_not_null(response, 'feed_publishers')
+        assert len(feeds) == 2
+        assert feeds[0].get('id') == 'here'
+        assert feeds[0].get('name') == 'here'
+        assert feeds[0].get('license') == 'Licence Ouverte / Open License'
+        assert feeds[0].get('url') == 'route.cit.api.here.com'
+
+        assert len(feeds) == 2
```
Comment (benoit-bst) [inline source/jormungandr/tests/here_distributed_routing_tests.py]: Check already exist line:352

**Review by benoit-bst — APPROVED** [review]

File: source/jormungandr/jormungandr/street_network/here.py [hunk 2]
```diff
@@ -42,7 +42,14 @@
 from copy import deepcopy
 from jormungandr.street_network.street_network import AbstractStreetNetworkService, StreetNetworkPathKey, \
     StreetNetworkPathType
+from jormungandr.ptref import FeedPublisher
 
+DEFAULT_HERE_FEED_PUBLISHER = {
+    'id': 'here',
+    'name': 'here',
+    'license': 'Licence Ouverte / Open License',
```
Comment (kinnou02) [inline source/jormungandr/jormungandr/street_network/here.py]: I am pretty sure it's not that

**Review by kinnou02 — DISMISSED** [review]

File: source/jormungandr/jormungandr/street_network/geovelo.py [hunk 3]
```diff
@@ -37,11 +37,20 @@
 from jormungandr.exceptions import TechnicalError, InvalidArguments, UnableToParse
 from jormungandr.street_network.street_network import AbstractStreetNetworkService, StreetNetworkPathKey
 from jormungandr.utils import get_pt_object_coord, is_url, decode_polyline
+from jormungandr.ptref import FeedPublisher
+
+DEFAULT_GEOVELO_FEED_PUBLISHER = {
+    'id': 'geovelo',
+    'name': 'geovelo',
+    'license': 'Licence Ouverte / Open License',
+    'url': 'route.cit.api.geovelo.com'
```
Comment (pbougue) [inline source/jormungandr/jormungandr/street_network/geovelo.py]: For the default url, I'd more see : http://about.geovelo.fr/cgu/
And the licence is proprietary, I'd say

File: source/jormungandr/jormungandr/street_network/here.py [hunk 4]
```diff
@@ -42,7 +42,14 @@
...
+    'url': 'route.cit.api.here.com'
```
Comment (pbougue) [inline source/jormungandr/jormungandr/street_network/here.py]: Same: licence proprietary and https://developer.here.com/terms-and-conditions is maybe a better url, but again, not sure

File: source/jormungandr/jormungandr/street_network/street_network.py [hunk 5]
```diff
@@ -90,6 +94,15 @@ def record_call(self, status, **kwargs):
         params.update(kwargs)
         new_relic.record_custom_event('streetnetwork', params)
 
+    def _add_feed_publisher(self, feed):
+        sr_feed = self.feed_publisher()
```
Comment (pbougue) [inline source/jormungandr/jormungandr/street_network/street_network.py]: probably `sn_feedpub` would be more explicit

File: source/jormungandr/jormungandr/street_network/here.py [hunk 6]
```diff
@@ -231,9 +239,13 @@ def direct_path(self, mode, pt_object_origin, pt_object_destination, fallback_ex
 
         self.log.debug('here response = {}'.format(json))
 
-        return self._read_response(json, pt_object_origin, pt_object_destination, mode,
+        resp = self._read_response(json, pt_object_origin, pt_object_destination, mode,
                                    fallback_extremity, request)
 
+        self._add_feed_publisher(resp)
```
Comment (pbougue) [inline source/jormungandr/jormungandr/street_network/here.py]: this should probably be handled at top level (so wrap direct_path call to do both : direct_path + feed)

File: source/jormungandr/jormungandr/street_network/street_network.py [hunk 7]
```diff
@@ -59,8 +59,15 @@ def get_street_network_routing_matrix(self, origins, destinations, street_networ
     def status(self):
         pass
 
+    def direct_path_with_fp(self, mode, pt_object_origin, pt_object_destination, fallback_extremity, request, direct_path_type):
```
Comment (pbougue) [inline source/jormungandr/jormungandr/street_network/street_network.py]: line length (same below)

File: source/jormungandr/jormungandr/street_network/kraken.py [hunk 8]
```diff
@@ -68,7 +68,7 @@ def _reverse_journeys(self, response):
             j.sections.sort(utils.SectionSorter())
         return response
 
-    def direct_path(self, mode, pt_object_origin, pt_object_destination, fallback_extremity, request, direct_path_type):
+    def _direct_path(self, mode, pt_object_origin, pt_object_destination, fallback_extremity, request, direct_path_type):
```
Comment (pbougue) [inline source/jormungandr/jormungandr/street_network/kraken.py:71]: technically it's more the implem of direct_path_with_fp() (it overloads it)
> ☑️ OK, as it's an abstract method, let's let it that way

File: source/jormungandr/jormungandr/street_network/kraken.py [hunk 9]
```diff
@@ -153,3 +153,7 @@ def make_path_key(self, mode, orig_uri, dest_uri, streetnetwork_path_type, perio
         direct path from A to B remains the same even the departure time are different (no realtime)
         """
         return StreetNetworkPathKey(mode, orig_uri, dest_uri, streetnetwork_path_type, None)
+
+    def feed_publisher(self):
```
Comment (pbougue) [inline source/jormungandr/jormungandr/street_network/kraken.py]: useless if we remove abstract

File: source/jormungandr/jormungandr/street_network/street_network.py [hunk 10]
```diff
@@ -79,6 +86,10 @@ def make_path_key(self, mode, orig_uri, dest_uri, streetnetwork_path_type, perio
         """
         pass
 
+    @abc.abstractmethod
```
Comment (pbougue) [inline source/jormungandr/jormungandr/street_network/street_network.py]: I'd remove that abstract and have it return None by default (any connector who wants to add fp would overload, but not all)

File: source/jormungandr/jormungandr/street_network/street_network.py [hunk 11]
```diff
@@ -90,6 +101,15 @@ def record_call(self, status, **kwargs):
         params.update(kwargs)
         new_relic.record_custom_event('streetnetwork', params)
 
+    def _add_feed_publisher(self, resp):
+        sr_feed = self.feed_publisher()
```
Comment (pbougue) [inline source/jormungandr/jormungandr/street_network/street_network.py]: typo: s`n`_feed I guess

File: source/jormungandr/jormungandr/street_network/valhalla.py [hunk 12]
```diff
@@ -317,3 +317,6 @@ def make_path_key(self, mode, orig_uri, dest_uri, streetnetwork_path_type, perio
         direct path from A to B remains the same even the departure time are different (no realtime)
         """
         return StreetNetworkPathKey(mode, orig_uri, dest_uri, streetnetwork_path_type, None)
+
+    def feed_publisher(self):
```
Comment (pbougue) [inline source/jormungandr/jormungandr/street_network/valhalla.py]: useless if we remove abstract

**Review by pbougue — APPROVED** [review]

**Pull request merged**